### PR TITLE
Pdp v2

### PR DESCRIPTION
**Pull request by RobertSamoilescu**
API improvement for the black-box model. Included the PDP computation in the code which allowed to remove the black-box wrapper.

### CODE REVIEW

File: doc/source/methods/PartialDependence.ipynb [hunk 1]
```diff
@@ -117,52 +90,64 @@
...
     "\n",
-    "- `response_method` - Specifies the prediction function to be used. For a classifier it specifies whether to use the `predict_proba` or the `decision_function`. For a regressor, the parameter is ignored. If set to `auto`, the `predict_proba` is tried first, and if not supported then it reverts to `decision_function`. Note that if `method='recursion'`, the prediction function always uses `decision_function`.\n",
+    "\n",
+    "- `response_method` - Specifies the prediction function to be used. For a classifier it specifies whether to use the `predict_proba` or the `decision_function` method. For a regressor, the parameter is ignored. If set to ``'auto'``, the `predict_proba` is tried first, and if not supported then it reverts to `decision_function`. If ``method='recursion'``, the prediction function always uses `decision_function`. Ignored for black-box models.\n",
```
Comment (jklaise) [inline doc/source/methods/PartialDependence.ipynb]: I would finish the first sentence with "...used for `sklearn` models."

File: doc/source/methods/PartialDependence.ipynb [hunk 2]
```diff
@@ -117,52 +90,64 @@
...
     "\n",
     "- `method` - The method used to calculate the partial dependence (i.e., the marginal effect one or two features have on the outcome of the predictor):\n",
     "    \n",
     "    * ``'auto'`` - uses ``'recursion'`` if the `predictor` supports it. Otherwise, uses the ``'brute'`` method.\n",
     "    \n",
-    "    * ``'brute'`` - supported for any black-box prediction model, but is more computationally intensive.\n",
+    "    * ``'brute'`` - supported for any `sklearn` and black-box prediction model, but is more computationally\n",
+    "    intensive.\n",
     "    \n",
     "    * ``'recursion'`` - a faster alternative only supported by some tree-based model. For a classifier, the \n",
```
Comment (jklaise) [inline doc/source/methods/PartialDependence.ipynb]: tree-based model -> tree-based `sklearn` models

File: alibi/explainers/partial_dependence.py [hunk 3]
```diff
@@ -158,16 +122,15 @@ def explain(self,  # type: ignore[override]
             `predict_proba` or the `decision_function` method. For a regressor, the parameter is ignored.
             If set to ``'auto'``, the `predict_proba` is tried first, and if not supported then it reverts to
             `decision_function`. If ``method='recursion'``, the prediction function always uses `decision_function`.
-            For a black-box model, the `response_method` should always be set to ``'auto'`` since the prediction
-            method is automatically inferred from the `predictor_kw` argument passed to
-            :py:meth:`alibi.explainers.partial_dependence.PartialDependence.__init__`.
+            Ignored for black-box models.
```
Comment (jklaise) [inline alibi/explainers/partial_dependence.py]: I would finish the first sentence with "...used for sklearn models."

File: alibi/explainers/partial_dependence.py [hunk 4]
```diff
@@ -388,18 +350,31 @@ def check_feature(f):
...
+        Parameters sanity checks for black-box models.
+
+        Parameters
+        ----------
+        method
+            See :py:meth:`alibi.explainers.partial_dependence.PartialDependence.explain` method.
+        """
+        if method != Method.BRUTE:
+            raise ValueError(f"method='{method}' is invalid. For black-box models "
+                             f"the method must be {Method.BRUTE}.")
```
Comment (jklaise) [inline alibi/explainers/partial_dependence.py:365]: Should this be `Method.BRUTE.value`?

File: alibi/explainers/partial_dependence.py [hunk 5]
```diff
@@ -489,68 +463,45 @@ def _partial_dependence(self,
...
         for f in features:  # type: ignore[union-attr]
-            # Note that we are using the _safe_indexing procedure implemented in sklearn to retrieve the column.
-            # This is because the _safe_indexing supports many data types such as sparse matrix representation,
-            # pandas dataframes etc. In case we would like to enhance alibi in the future to support other
-            # data types than np.ndarrays, this can be useful.
-            f_indices = np.asarray(_get_column_indices(X, f), dtype=np.int32, order='C').ravel()
-            X_f = _safe_indexing(X, f_indices, axis=1)
-
+            # extract column
+            X_f = X[:, f]
```
Comment (jklaise) [inline alibi/explainers/partial_dependence.py:467]: Perhaps it's worth leaving a `TODO` comment here with the option of using `_safe_indexing` in the future to support more input types?

File: alibi/explainers/partial_dependence.py [hunk 6]
```diff
@@ -489,68 +463,45 @@ def _partial_dependence(self,
...
                 # construct grid for feature f. Note that for categorical features we pass the
                 # grid resolution to be infinity because otherwise we risk to apply `linspace` to
                 # categorical values, which does not make sense.
-                _, values_f = _grid_from_X(X_f,
-                                           percentiles=percentiles,
-                                           grid_resolution=grid_resolution if self._is_numerical(f) else np.inf)
+                values_f = self._grid_from_X(X=X_f.reshape(-1, 1),
+                                             percentiles=percentiles,
+                                             grid_resolution=grid_resolution  # type: ignore[arg-type]
+                                             if self._is_numerical(f) else np.inf)
```
Comment (jklaise) [inline alibi/explainers/partial_dependence.py]: Does this code actually work with the line being split...?

File: alibi/explainers/tests/test_partial_dependence.py [hunk 7]
```diff
@@ -284,63 +281,37 @@ def test_explanation_numerical_shapes(rf_classifier, iris_data, grid_resolution,
...
 @pytest.mark.parametrize('kind', ['average', 'individual', 'both'])
 @pytest.mark.parametrize('features', [
     [0, 1, 2],
     [(0, 1), (0, 2), (1, 2)],
     [0, 1, (0, 1)]
 ])
-def test_classification_wrapper(lr_classifier, iris_data, response_method, kind, features):
-    """ Test the black-box wrapper for a classification function. """
-    X_train, y_train = iris_data['X_train'], iris_data['y_train']
+def test_balckbox_classification(lr_classifier, iris_data, kind, features):
```
Comment (jklaise) [inline alibi/explainers/tests/test_partial_dependence.py]: balckbox -> blackbox

**Review by jklaise — APPROVED** [review]
Looks good! Left a few minor comments. Much nicer to not have the `predictor_kw` argument and agree on vendoring the `sklearn` private functions into `alibi` so we have full control over them in the future.

File: alibi/explainers/partial_dependence.py (hunk 4 above)
Comment (RobertSamoilescu) [reply alibi/explainers/partial_dependence.py:365]: not necessarily. see `_sklearn_params_sanity_checks` where it works just fine.

**Pull request merged**